R/cnd-handlers.R

```diff
@@ -1,8 +1,136 @@
+#'     catching site instead.
+#'
+#'     ```
+#'     # Recover with the value 0
+#'     try_catch(1 + "", error = function(cnd) 0)
+#'     ```
+#'
+#' -   **Rethrow conditions**, e.g. using `abort(msg, parent = cnd)`.
+#'     See the `parent` argument of [abort()]. This is typically done to
+#'     add information to low-level errors about the high-level context
+#'     in which they occurred.
+#'
+#'     ```
+#'     try_catch(1 + "", error = function(cnd) abort("Failed.", parent = cnd))
+#'     ```
+#'
+#' -   **Inspect conditions**, for instance to log data about warnings
+#'     or errors. In this case, the handler must return the [zap()]
+#'     sentinel to instruct `tryCatch()` to ignore (or zap) that
+#'     particular handler. The next matching handler is called if any,
+#'     and errors bubble up to the user if no handler remains.
+#'
+#'     ```
+#'     log <- NULL
+#'     try_catch(1 + "", error = function(cnd) {
+#'       log <<- cnd
+#'       zap()
+#'     })
+#'     ```
+#'
+#' @param expr An R expression.
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> Named condition
+#'   handlers. The names specify the condition class for which a
+#'   handler will be called.
+#'
+#'   Handler functions must have at least two arguments. By
+#'   convention, the second argument should be named `throw`.
+#'
+#' @section Stack overflows:
+#'
+#' A stack overflow occurs when a program keeps adding to itself until
+#' the stack memory (whose size is very limited unlike heap memory) is
+#' exhausted.
+#'
+#' ```
+#' # A function that calls itself indefinitely causes stack overflows
+#' f <- function() f()
+#' f()
+#' #> Error: C stack usage  9525680 is too close to the limit
+#' ```
+#'
+#' Because `try_catch()` preserves as much of the running program as
+#' possible in order to produce informative backtraces on error
+#' rethrows, it is unable to deal with stack overflows. When an
+#' overflow occurs, R needs to unwind as much of the program is
+#' possible and can't afford to run any more R code, such as in an
+#' error handler.
+#'
+#' This usually isn't a problem for error rethrowing or logging but
+#' might make your program more brittle in case of error recovery. In
+#' that case, capture errors with [base::tryCatch()] instead of
+#' `try_catch()`.
+#'
+#' @section Comparison with `tryCatch()`:
+#'
+#' `try_catch()` generalises `tryCatch()` and `withCallingHandlers()`
+#' in a single function. It reproduces the behaviour of both calling
+#' and exiting handlers depending the on the return value of the
+#' handler. If the handler returns the [zap()] sentinel, it is taken
+#' as a calling handler that declines to recover from a condition.
+#' Otherwise, it is taken as an exiting handler which returns a value
+#' from the catching site.
+#'
+#' There is an important difference between `tryCatch()` and
+#' `try_catch()`. The latter doesn't catch condition objects, instead
+#' it catches recovery values returned by the handler. The main
```

Comment:
I don't understand what this means. Could you reword it to say something like "preserves the call stack"?